### PR TITLE
178837538 opt out form

### PR DESCRIPTION
**Pull request by kbiala**
#### What's this PR do?
- Removes all references to privacy apps, and replaces them with simple opt-out form.

#### How should this be manually tested?
- Insall the project (download master.key from keybase - you will need it for captcha keys)
- Use 'Opt Out' link in footer, play with the form, check emails under /letter_opener
- Log in as admin (see seeds) and review the admin panel available under 'Opt Out requests' in footer
- Confirm that there're no references to donotsell website any more
 
#### What are the relevant tickets?
- https://www.pivotaltracker.com/n/projects/2177859/stories/178837538

#### Task completed checklist:
- [x] Is there appropriate test coverage?
- [x] Did I take into consideration possible security vulnerabilities?
- [x] Are the controllers secure?
- [x] Was every important matter documented?

#### Things to watch out for
- [N] Does this PR have any migrations?
- [N] Does this PR add new dependencies?

### CODE REVIEW

File: app/controllers/opt_out_requests_controller.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
+class OptOutRequestsController < ApplicationController
+  def index
+    @opt_out_requests = OptOutRequest.all
```
Comment (psuchta) [inline app/controllers/opt_out_requests_controller.rb]: We can delete this line

File: app/controllers/opt_out_requests_controller.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+          referer: request.referrer,
+          device: browser.device.name,
+          platform: browser.platform.name,
+          bot: browser.bot?,
+          bot_reason: browser.bot.why?
+        }
+      )
+    )
+    if verify_recaptcha(model: @opt_out_request) && @opt_out_request.save
+      AdminMailer.new_opt_out_request_notice(@opt_out_request.id).deliver_now unless browser.bot?
```
Comment (psuchta) [inline app/controllers/opt_out_requests_controller.rb:38]: Do we expect bots to go so deep into the code ?

File: db/migrate/20210805135246_create_opt_out_requests.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
+class CreateOptOutRequests < ActiveRecord::Migration[6.1]
+  def change
+    create_table :opt_out_requests do |t|
+      t.integer :request_type
+      t.text   :meta_data
```
Comment (psuchta) [inline db/migrate/20210805135246_create_opt_out_requests.rb]: Meta data has a structure of a JSON field, if we want to filter records by this column we should consider changing its type to JSON I think.

File: Gemfile [hunk 4]
```diff
@@ -105,3 +107,13 @@ gem 'aws-sdk-s3', require: false
...
+
+# https://github.com/ambethia/recaptcha
+gem 'recaptcha'
+
+gem 'browser'
+
+gem 'kaminari'
+
+gem 'wkhtmltopdf-binary'
+gem 'wicked_pdf'
```
Comment (psuchta) [inline Gemfile]: In some changed files, Github is showing that an empty line at the end of the file is missing, if it's important please correct that

File: app/views/contact_requests/index.html.erb [hunk 5]
```diff
@@ -25,3 +26,5 @@
     <% end%>
   </tbody>
 </table>
+
+<%= paginate @contact_requests %>
```
Comment (psuchta) [inline app/views/contact_requests/index.html.erb:30]: Please add padding between numbers in the pagination form

File: test/models/opt_out_request_test.rb [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  setup do
+    @opt_out_request = build :opt_out_request
+  end
+
+  describe '.create' do
+    test 'opt-out request without type is not created' do
+      @opt_out_request.request_type = ''
+      assert_no_difference('OptOutRequest.count') do
+        @opt_out_request.save
+      end
```
Comment (psuchta) [inline test/models/opt_out_request_test.rb]: Worth considering using `refute @opt_out_request.valid?` in these tests

File: test/models/user_test.rb [hunk 7]
```diff
@@ -9,21 +9,25 @@ class UserTest < ActiveSupport::TestCase
...
   end
 
-  test 'correctly creates user with username, encrypts email, creates hash' do
-    assert_difference('User.count') do
-      User.create(@user_params)
-    end
+  describe '.create' do
+    test 'correctly creates user with username, encrypts email, creates hash' do
+      assert_difference('User.count') do
+        User.create(@user_params)
```
Comment (psuchta) [inline test/models/user_test.rb]: We can use factories here

**Review by psuchta — CHANGES_REQUESTED** [review]

File: app/controllers/opt_out_requests_controller.rb (hunk 2 above)
Comment (kbiala) [reply app/controllers/opt_out_requests_controller.rb:38]: Might happen. Captcha is not perfect, so it wouldn't hurt to add another layer to at least prevent from sending spam.

File: db/migrate/20210805135246_create_opt_out_requests.rb (hunk 3 above)
Comment (kbiala) [reply db/migrate/20210805135246_create_opt_out_requests.rb]: Good point

File: app/controllers/opt_out_requests_controller.rb (hunk 2 above)
Comment (psuchta) [reply app/controllers/opt_out_requests_controller.rb:38]: how about adding additional condition in `if` statement?
Smething like:
```
!browser.bot? && verify_recaptcha(model: @opt_out_request) && @opt_out_request.save
```

Comment (kbiala) [reply app/controllers/opt_out_requests_controller.rb:38]: I prefer to keep these records in db for now - we can see why record was defined as sent by bot (bot_reason property), and then we can modify configuration if needed.

**Review by psuchta — APPROVED** [review]

**Pull request merged**